### PR TITLE
feat(ucm): Add new completion pattern for upgrade

### PR DESCRIPTION
**Pull request by sixfourtwelve**
**Choose your PR title well:** Your pull request title is what's used to create release notes, so please make it descriptive of the change itself, which may be different from the initial motivation to make the change.

## Overview

What does this change accomplish and why?
i.e. How does it change the user experience?
i.e. What was the old behavior/API and what is the new behavior/API?

Feel free to include "before and after" examples if appropriate. (You can copy/paste screenshots directly into this editor.)

If relevant, which Github issues does it close? (See [closing-issues-using-keywords](https://help.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords).)

## Implementation notes

How does it accomplish it, in broad strokes? i.e. How does it change the Haskell codebase?

## Interesting/controversial decisions

Include anything that you thought twice about, debated, chose arbitrarily, etc.
What could have been done differently, but wasn't? And why?

## Test coverage

Have you included tests (which could be a transcript) for this change, or is it somehow covered by existing tests?

Would you recommend improving the test coverage (either as part of this PR or as a separate issue) or do you think it’s adequate?

If you only tested by hand, because that's all that's practical to do for this change, mention that.

## Loose ends

Link to related issues that address things you didn't get to. Stuff you encountered on the way and decided not to include in this PR.

### CODE REVIEW

**Comment by sixfourtwelve** [comment]
note: add transcript tests

**Comment by aryairani** [comment]
Hi @sixfourtwelve — 
For this one, we want it to complete just the names of the children under `<branch-root>.lib.`

**Comment by aryairani** [comment]
I converted it to a draft while you're still working on it. Feel free to ask in #development on Slack if you have any questions!

File: unison-cli/src/Unison/CommandLine/InputPatterns.hs [hunk 1]
```diff
@@ -2801,7 +2801,7 @@ upgrade =
     { patternName = "upgrade",
       aliases = [],
       visibility = I.Visible,
-      argTypes = [],
+      argTypes = [(ZeroPlus, upgradeArg)],
```
Comment (aryairani) [inline unison-cli/src/Unison/CommandLine/InputPatterns.hs:2804]: This should be two mandatory arguments; the old and the new libraries.

File: unison-cli/src/Unison/CommandLine/Completion.hs [hunk 2]
```diff
@@ -294,7 +296,16 @@ prefixCompletePatch ::
...
   Sqlite.Transaction [Line.Completion]
-prefixCompletePatch = completeWithinNamespace (NESet.singleton PatchCompletion)
+prefixCompletePatch = completeWithinNamespace (NESet.singleton UpgradeCompletion)
+
+-- | Completes a uograde argument by prefix-matching against the query.
+prefixCompleteUpgrade ::
+  String ->
+  Path.Absolute -> -- Current path
+  Sqlite.Transaction [Line.Completion]
+prefixCompleteUpgrade = completeWithinNamespace (NESet.singleton PatchCompletion)
```
Comment (aryairani) [inline unison-cli/src/Unison/CommandLine/Completion.hs]: It looks like the arguments for Upgrade and Patch got swapped here?

And then what happenes when `completeWithinNamespace` gets called with `UpgradeCompletion`? I haven't found the spot yet 😅

File: unison-cli/src/Unison/CommandLine/InputPatterns.hs [hunk 3]
```diff
@@ -3020,6 +3020,14 @@ patchArg =
       globTargets = Set.fromList []
     }
 
+upgradeArg :: ArgumentType
+upgradeArg =
+  ArgumentType
+    { typeName = "upgrade",
+      suggestions = \q cb _http p -> Codebase.runTransaction cb (prefixCompleteUpgrade q p),
+      globTargets = Set.fromList [Globbing.Term, Globbing.Type]
```
Comment (aryairani) [inline unison-cli/src/Unison/CommandLine/InputPatterns.hs:3028]: We probably don't want `Globbing.Term` or `Globbing.Type` here, we are looking for the top-level paths under `lib`.

**Pull request closed without merging**